crates/ruff_formatter/src/builders.rs

```diff
@@ -361,31 +360,62 @@ impl std::fmt::Debug for DynamicText<'_> {
     }
 }
 
-/// Creates a text from a dynamic string and a range of the input source
-pub fn static_text_slice(text: Rc<str>, range: TextRange) -> StaticTextSlice {
-    debug_assert_no_newlines(&text[range]);
+/// Emits a text as it is written in the source document. Optimized to avoid allocations.
+pub const fn source_text_slice(
+    range: TextRange,
+    new_lines: ContainsNewLines,
+) -> SourceTextSliceBuilder {
+    SourceTextSliceBuilder { range, new_lines }
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum ContainsNewLines {
```

Comment:
Sorry, nit, but I think this should be `ContainsNewlines` (we treat it as a single word elsewhere in this crate).